README.md
# @guoyunhe/create-cli

Initialize a Node.js command line tool project

## Get Started

### Create a new project

```bash
npm create @guoyunhe/cli my-cli
```

### Initialize an existing project

```bash
cd my-cli
npm init @guoyunhe/cli
```

### Project structure

```bash
├── dist # Build output
│   ├── index.js # CJS API entry
│   ├── index.mjs # ESM API entry
│   ├── index.d.ts # TypeScript declaration
│   ├── my-cli.js # CJS CLI script
│   └── my-cli.mjs # ESM CLI script
├── src # Source code
│   ├── bin # CLI scripts
│   │   └── my-cli.ts
│   ├── index.test.ts # API unit test
│   └── index.ts # API entry (add all exports here)
├── .editorconfig
├── .gitignore
├── CHANGELOG.md
├── package.json
├── README.md
└── tsconfig.json
```

### Package scripts

```bash
# Build output
npm run build
# Build output in watch mode
npm run watch
# Format source code
npm run format
# Check lint issues
npm run lint
# Run unit tests (support all jest command options)
npm test
# Run unit tests in watch mode
npm test -- --watch
# Update unit test snapshots
npm test -- -u
```

## Advanced Options

### Initial package version

1.0.0 by default.

```bash
npm create @guoyunhe/cli my-cli --package-version 0.1.0
```

### Add multiple bin

Let's say, you want to add a new bin called `health-check`.

First, create `src/bin/health-check.ts`, with shebang `#!/usr/bin/env node`:

```ts
#!/usr/bin/env node

console.log('Doing health check...');

// Add your code...
```

Then, add bin entry to your `package.json`:

```json
{
  "bin": {
    "health-check": "dist/health-check.js"
  }
}
```


CHANGELOG.md
# Changelog

## 1.8.0 - 2023-06-24

- Updated to nodejs 20
- Updated to typescript 5

## 1.7.0 - 2023-02-18

- Changed `package.json` template:
  - Updated `commander` to 10.0, requires Node.js 14+
- Updated `commander` to 10.0, requires Node.js 14+
- Changed `--node-version` options, drop Node.js `12` support

## 1.6.0 - 2023-02-18

- Changed `package.json` template:
  - Updated `@guoyunhe/node-scripts` to 2.0
- Changed `bin` template:
  - Removed unneeded null check for `PACKAGE_VERSION`

## 1.5.1 - 2023-02-11

- Fixed bin path

## 1.5.0 - 2023-02-11

- Updated `@guoyunhe/node-scripts` to 2.0
- Splitted `package.json` template into `package-defaults.json` and `package-overrides.json`
- Changed to not override `src/index.ts`, `README.md`, `CHANGELOG.md`
- Simplified `.gitignore` template

## 1.4.1 - 2023-01-26

- Change Prettier `printWidth` to `100`
- Fix VSCode `editor.rulers` settings

## 1.4.0 - 2023-01-22

- Initialize `description` and `keywords`
- Use `package-json-from-git` to initialize `repository`, `homepage`, `bugs` and `author` fields in `package.json` from Git data

## 1.3.0 - 2022-12-17

- Add `typings` attribute to `package.json`

## 1.2.0 - 2022-12-17

- Rename `cli-scripts` to `node-scripts`

## 1.1.0 - 2022-12-17

- Format `package.json` with `sort-package-json`

## 1.0.0 - 2022-12-16

- Support create basic cli project
- Support `--package-version` option
- Support `--node-version` option
- Support `--strict` option


src/index.spec.ts
import { createProject } from '.';

describe('exports', () => {
  it('exports createProject', () => {
    expect(typeof createProject).toBe('function');
  });
});


src/index.ts
import { outputFile, outputJSON, pathExists, readJSON } from 'fs-extra';
import merge from 'merge';
import { getPackageJsonFromGit } from 'package-json-from-git';
import { basename, join } from 'path';
import sortPackageJson from 'sort-package-json';
import binTs from './template/bin.txt';
import changelog from './template/changelog.txt';
import editorconfig from './template/editorconfig.txt';
import gitignore from './template/gitignore.txt';
import indexTestTs from './template/index.test.txt';
import indexTs from './template/index.txt';
import packageJsonDefaults from './template/package-defaults.json';
import packageJsonOverrides from './template/package-overrides.json';
import readme from './template/readme.txt';
import tsconfigJson from './template/tsconfig.json';
import vscodeExtensions from './template/vscode-extensions.json';
import vscodeSettings from './template/vscode-settings.json';

export interface CreateProjectOptions {
  // Initial version number, 1.0.0 by default
  packageVersion?: string;
}

export async function createProject(
  project: string | null,
  { packageVersion }: CreateProjectOptions
) {
  const projectFullPath = project ? join(process.cwd(), project) : process.cwd();

  console.log('Project full path: ', projectFullPath);

  // .vscode/settings.json
  const vscodeSettingsPath = join(projectFullPath, '.vscode', 'settings.json');
  outputJSON(vscodeSettingsPath, vscodeSettings, { spaces: 2 });

  // .vscode/extensions.json
  const vscodeExtensionsPath = join(projectFullPath, '.vscode', 'extensions.json');
  outputJSON(vscodeExtensionsPath, vscodeExtensions, { spaces: 2 });

  // .editorconfig
  const editorConfigPath = join(projectFullPath, '.editorconfig');
  outputFile(editorConfigPath, editorconfig);

  // .gitignore
  const gitignorePath = join(projectFullPath, '.gitignore');
  outputFile(gitignorePath, gitignore);

  // tsconfig.json
  const tsconfigJsonPath = join(projectFullPath, 'tsconfig.json');
  outputJSON(tsconfigJsonPath, tsconfigJson, { spaces: 2 });

  // package.json
  const packageJsonPath = join(projectFullPath, 'package.json');
  let oldPackageJson: any = {};
  if (await pathExists(packageJsonPath)) {
    oldPackageJson = await readJSON(packageJsonPath, { throws: false });
  }
  const newPackageJson: any = await getPackageJsonFromGit();
  merge.recursive(newPackageJson, packageJsonDefaults, oldPackageJson, packageJsonOverrides);
  newPackageJson.name ||= basename(projectFullPath);
  newPackageJson.version = packageVersion || newPackageJson.version || '1.0.0';
  const binName = basename(newPackageJson.name);
  newPackageJson.bin = {
    [binName]: `dist/${binName}.js`,
  };
  outputJSON(packageJsonPath, sortPackageJson(newPackageJson), { spaces: 2 });

  // CHANGELOG.md
  const changelogPath = join(projectFullPath, 'CHANGELOG.md');
  if (!(await pathExists(changelogPath))) {
    const date = new Date().toISOString().substring(0, 10);
    const newChangelog = changelog
      .replaceAll('%date%', date)
      .replaceAll('%version%', newPackageJson.version);
    outputFile(changelogPath, newChangelog);
  }

  // README.md
  const readmePath = join(projectFullPath, 'README.md');
  if (!(await pathExists(readmePath))) {
    const newReadme = readme.replaceAll('%name%', newPackageJson.name);
    outputFile(readmePath, newReadme);
  }

  // src/index.ts
  const indexTsPath = join(projectFullPath, 'src', 'index.ts');
  if (!(await pathExists(indexTsPath))) {
    outputFile(indexTsPath, indexTs);

    // src/index.test.ts
    const indexTestTsPath = join(projectFullPath, 'src', 'index.test.ts');
    outputFile(indexTestTsPath, indexTestTs);
  }

  // src/bin/binName.ts
  const binTsPath = join(projectFullPath, 'src', 'bin', binName + '.ts');
  if (!(await pathExists(binTsPath))) {
    const newBinTs = binTs.replaceAll('%binName%', binName);
    outputFile(binTsPath, newBinTs);
  }
}


src/lib/createProject.ts
import { outputFile, outputJSON, pathExists, readJSON } from 'fs-extra';
import merge from 'merge';
import { getPackageJsonFromGit } from 'package-json-from-git';
import { basename, join } from 'path';
import sortPackageJson from 'sort-package-json';
import binTs from './template/bin.txt';
import changelog from './template/changelog.txt';
import editorconfig from './template/editorconfig.txt';
import gitignore from './template/gitignore.txt';
import indexTestTs from './template/index.test.txt';
import indexTs from './template/index.txt';
import packageJson from './template/package.json';
import readme from './template/readme.txt';
import tsconfigJson from './template/tsconfig.json';
import vscodeExtensions from './template/vscode-extensions.json';
import vscodeSettings from './template/vscode-settings.json';

export interface CreateProjectOptions {
  // Initial version number, 1.0.0 by default
  packageVersion?: string;
  // Minimum supported Node.js version, 16 by default
  nodeVersion?: '12' | '14' | '16' | '18';
  // Use strict TypeScript configuration
  strict: boolean;
}

export async function createProject(
  project: string | null,
  { packageVersion, nodeVersion = '16', strict }: CreateProjectOptions
) {
  if (!['12', '14', '16', '18'].includes(nodeVersion)) {
    throw new Error('--node-version must be one of the following: 12, 14, 16, 18');
  }

  const projectFullPath = project ? join(process.cwd(), project) : process.cwd();

  console.log('Project full path: ', projectFullPath);

  // .vscode/settings.json
  const vscodeSettingsPath = join(projectFullPath, '.vscode', 'settings.json');
  outputJSON(vscodeSettingsPath, vscodeSettings, { spaces: 2 });

  // .vscode/extensions.json
  const vscodeExtensionsPath = join(projectFullPath, '.vscode', 'extensions.json');
  outputJSON(vscodeExtensionsPath, vscodeExtensions, { spaces: 2 });

  // .editorconfig
  const editorConfigPath = join(projectFullPath, '.editorconfig');
  outputFile(editorConfigPath, editorconfig);

  // .gitignore
  const gitignorePath = join(projectFullPath, '.gitignore');
  outputFile(gitignorePath, gitignore);

  // tsconfig.json
  const tsconfigJsonPath = join(projectFullPath, 'tsconfig.json');
  let tsconfigPreset: string;
  if (['16', '18'].includes(nodeVersion) && strict) {
    tsconfigPreset = `@tsconfig/node${nodeVersion}-strictest`;
  } else {
    tsconfigPreset = `@tsconfig/node${nodeVersion}`;
  }
  const newTsconfigJson = { extends: tsconfigPreset + '/tsconfig.json', ...tsconfigJson };
  outputJSON(tsconfigJsonPath, newTsconfigJson, { spaces: 2 });

  // package.json
  const packageJsonPath = join(projectFullPath, 'package.json');
  let oldPackageJson: any = {};
  if (await pathExists(packageJsonPath)) {
    oldPackageJson = await readJSON(packageJsonPath, { throws: false });
  }
  const newPackageJson: any = await getPackageJsonFromGit();
  merge.recursive(newPackageJson, oldPackageJson, packageJson);
  newPackageJson.name ||= basename(projectFullPath);
  newPackageJson.version = packageVersion || newPackageJson.version || '1.0.0';
  newPackageJson.devDependencies[tsconfigPreset] = '^1.0.0';
  newPackageJson.devDependencies['@types/node'] = `^${nodeVersion}.0.0`;
  const binName = basename(newPackageJson.name);
  newPackageJson.bin = {
    [binName]: `dist/cjs/bin/${binName}.js`,
  };
  outputJSON(packageJsonPath, sortPackageJson(newPackageJson), { spaces: 2 });

  // CHANGELOG.md
  const changelogPath = join(projectFullPath, 'CHANGELOG.md');
  const date = new Date().toISOString().substring(0, 10);
  const newChangelog = changelog.replaceAll('%date%', date).replaceAll('%version%', newPackageJson.version);
  outputFile(changelogPath, newChangelog);

  // README.md
  const readmePath = join(projectFullPath, 'README.md');
  const newReadme = readme.replaceAll('%name%', newPackageJson.name).replaceAll('%binName%', binName);
  outputFile(readmePath, newReadme);

  // src/bin/binName.ts
  const binTsPath = join(projectFullPath, 'src', 'bin', binName + '.ts');
  const newBinTs = binTs.replaceAll('%binName%', binName);
  outputFile(binTsPath, newBinTs);

  // src/index.ts
  const indexTsPath = join(projectFullPath, 'src', 'index.ts');
  outputFile(indexTsPath, indexTs);

  // src/index.test.ts
  const indexTestTsPath = join(projectFullPath, 'src', 'index.test.ts');
  outputFile(indexTestTsPath, indexTestTs);
}
